src/XrdVoms/XrdVomsMapfile.cc

```diff
@@ -0,0 +1,359 @@
+/* along with XRootD in a file called COPYING.LESSER (LGPL license) and file  */
+/* COPYING (GPL license).  If not, see <http://www.gnu.org/licenses/>.        */
+/*                                                                            */
+/* The copyright holder's institutional names and contributor's names may not */
+/* be used to endorse or promote products derived from this software without  */
+/* specific prior written permission of the institution or contributor.       */
+/******************************************************************************/
+
+#include "XrdVoms/XrdVomsMapfile.hh"
+
+#include "XrdOuc/XrdOucEnv.hh"
+#include "XrdOuc/XrdOucString.hh"
+#include "XrdOuc/XrdOucStream.hh"
+#include "XrdSec/XrdSecEntity.hh"
+#include "XrdSys/XrdSysError.hh"
+
+#include <memory>
+#include <fstream>
+#include <sstream>
+#include <vector>
+#include <string>
+#include <fcntl.h>
+
+
+bool XrdVomsMapfile::tried_configure = false;
+std::unique_ptr<XrdVomsMapfile> XrdVomsMapfile::mapper;
+
+namespace {
+
+std::string
+PathToString(const std::vector<std::string> &path)
+{
+    if (path.empty()) {return "/";}
+    std::stringstream ss;
+    for (const auto &entry : path) {
+        ss << "/" << entry;
+    }
+
+    return ss.str();
+}
+
+}
+
+
+XrdVomsMapfile::XrdVomsMapfile(XrdSysError *erp, XrdHttpSecXtractor *xrdvoms,
+    const std::string &mapfile)
+    : m_mapfile(mapfile), m_xrdvoms(xrdvoms), m_edest(erp)
+{
+    m_last_update.store(0, std::memory_order_relaxed);
+}
+
+
+bool
+XrdVomsMapfile::ParseMapfile(const std::string &mapfile)
+{
+    std::ifstream fstr(mapfile);
+    if (!fstr.is_open()) {
+        m_edest->Emsg("ParseMapfile", "Failed to open file", mapfile.c_str(), strerror(errno));
+        return false;
+    }
+    std::shared_ptr<std::vector<MapfileEntry>> entries(new std::vector<MapfileEntry>());
+    for (std::string line; std::getline(fstr, line); ) {
+        MapfileEntry entry;
+        if (ParseLine(line, entry.m_path, entry.m_target) && !entry.m_path.empty()) {
+            if (m_edest->getMsgMask() & LogMask::Debug) {
+                m_edest->Log(LogMask::Debug, "ParseMapfile", PathToString(entry.m_path).c_str(), "->", entry.m_target.c_str());
+            }
+            entries->emplace_back(entry);
+        }
+    }
+    m_entries = entries;
+    return true;
+}
+
+
+bool
+XrdVomsMapfile::Reconfigure() {
+    auto now = time(NULL);
+    auto retval = true;
+    std::stringstream ss;
+    ss << "Last update " << m_last_update.load(std::memory_order_relaxed) << ", " << now;
+    m_edest->Log(LogMask::Debug, "VOMS Mapfile", ss.str().c_str());
+    if (now > m_last_update.load(std::memory_order_relaxed) + 30) {
+        retval = ParseMapfile(m_mapfile);
+        m_last_update.store(now, std::memory_order_relaxed);
+    }
+    return retval;
+}
+
+
+bool
+XrdVomsMapfile::ParseLine(const std::string &line, std::vector<std::string> &entry, std::string &target)
```

Comment:
We will need documentation on the format of the mapfile as well as the rules that are used for the actual mapping. Can you provide such a document so that it can be included in the security reference.

Reply:
Sure.  How about I also add it to a module-specific README?  The parsing rules are taken from here:

https://argus-documentation.readthedocs.io/en/stable/misc/grid_map_file.html

(except I will decline to support generic 8-bit ASCII or generic unicode escape sequences as those are not permitted in IGTF... would rather someone really convince me unicode support in DNs is a good idea!)

I wasn't able to find a document for the wildcard / matching rules so I took them from the existing implementation in order to allow sites to reuse their existing voms-mapfiles.  Would be good to have them written down.

Reply:
A READ_ME file is OK and I suppose I will have to convert it to reference text as people defining the map file generally do not install the source. Pointing then to the argus page won't help either because, well as you said, you don't support all of it and completely agree with you we don't need to support all of it.

Reply:
The README file is added in the latest set of commits.  Resolving.